libparsec/crates/client/src/client/recovery_device.rs

```diff
@@ -0,0 +1,373 @@
+
+use std::{
+    path::{Path, PathBuf},
+    sync::Arc,
+};
+
+use libparsec_client_connection::{AuthenticatedCmds, ConnectionError, ProxyConfig};
+use libparsec_platform_device_loader::{
+    get_default_key_file, save_device, PlatformExportRecoveryDeviceError,
+    PlatformImportRecoveryDeviceError, SaveDeviceError,
+};
+use libparsec_platform_storage::certificates::{GetCertificateError, PerTopicLastTimestamps};
+use libparsec_protocol::authenticated_cmds::v4::device_create;
+use libparsec_types::{
+    AvailableDevice, DeviceLabel, DeviceSaveStrategy, LocalDevice, SecretKeyPassphrase,
+};
+
+use crate::{
+    greater_timestamp, CertifPollServerError, CertificateBasedActionOutcome,
+    GreaterTimestampOffset, InvalidCertificateError,
+};
+use libparsec_types::prelude::*;
+
+use super::Client;
+
+pub async fn export_recovery_device(
+    client: &Client,
+    device_label: DeviceLabel,
+) -> Result<(SecretKeyPassphrase, Vec<u8>), ExportRecoveryDeviceError> {
+    let (passphrase, data, recovery_device) =
+        libparsec_platform_device_loader::export_recovery_device(&client.device, device_label)
+            .await?;
+
+    // save recovery device
+    let outcome =
+        register_new_device(client.cmds.clone(), recovery_device, client.device.clone()).await?;
+
+    let latest_known_timestamps = match outcome {
+        CertificateBasedActionOutcome::LocalIdempotent => return Ok((passphrase, data)), // not sure how it could happen though
+        CertificateBasedActionOutcome::Uploaded {
+            certificate_timestamp,
+        }
+        | CertificateBasedActionOutcome::RemoteIdempotent {
+            certificate_timestamp,
+        } => PerTopicLastTimestamps::new_for_common_and_shamir(
+            certificate_timestamp,
+            certificate_timestamp,
+        ),
+    };
+    client
+        .certificates_ops
+        .poll_server_for_new_certificates(Some(&latest_known_timestamps))
+        .await
+        .map_err(|e| match e {
+            CertifPollServerError::Stopped => ExportRecoveryDeviceError::Stopped,
+            CertifPollServerError::Offline => ExportRecoveryDeviceError::Offline,
+            CertifPollServerError::InvalidCertificate(err) => {
+                ExportRecoveryDeviceError::InvalidCertificate(err)
+            }
+            CertifPollServerError::Internal(err) => err
+                .context("Cannot poll server for new certificates")
+                .into(),
+        })?;
+    Ok((passphrase, data))
+}
+
+pub async fn create_device_from_recovery(
+    cmds: Arc<AuthenticatedCmds>,
+    recovery_device: &LocalDevice,
+    device_label: &DeviceLabel,
+    save_strategy: DeviceSaveStrategy,
+    config_dir: PathBuf,
+) -> Result<AvailableDevice, ImportRecoveryDeviceError> {
+    let new_device =
+        LocalDevice::from_existing_device_for_user(recovery_device, device_label.clone());
+    // save recovery device
+    // TODO: save after upload
```

Comment:
Is there an issue to create about this ?

What is the tradeof here ? I guess it's something similar to what we have when we create a device during enrollment (i.e. if our machine goes down at the wrong time we have the choice between losing the new device, or having in our client a device the server doesn't know about [^1])

In the end I think we should follow the same strategy in both places  (so saving the device *after* we have uploaded it to the server)

[^1]: Of course we can have the best of both world with a more complex approach based on saving a temporary type that represent the fact the device might or might have been uploaded on the server... but it's too much complexity for a corner case, at least for the moment ;-)

Reply:
see #8810